Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp

```diff
@@ -83,4 +103,58 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+{
+    ASSERT(sessionStorageQuota() != WebCore::StorageMap::noQuota);
+
+    auto& webPage = WebPage::fromCorePage(page);
+
+    // The identifier of a session storage namespace is the WebPageProxyIdentifier. It is possible we have several WebPage objects in a single process for the same
+    // WebPageProxyIdentifier and these need to share the same namespace instance so we know where to route the IPC to.
+    auto namespacesIt = m_sessionStorageNamespaces.find(webPage.sessionStorageNamespaceIdentifier());
+    if (namespacesIt == m_sessionStorageNamespaces.end()) {
+        if (shouldCreate == ShouldCreateNamespace::No)
+            return nullptr;
+        namespacesIt = m_sessionStorageNamespaces.set(webPage.sessionStorageNamespaceIdentifier(), SessionStorageNamespaces { }).iterator;
+    }
+
+    auto& sessionStorageNamespacesMap = namespacesIt->value.map;
+    auto it = sessionStorageNamespacesMap.find(topLevelOrigin.data());
+    if (it == sessionStorageNamespacesMap.end()) {
+        if (shouldCreate == ShouldCreateNamespace::No)
+            return nullptr;
+        auto sessionStorageNamespace = StorageNamespaceImpl::createSessionStorageNamespace(webPage.sessionStorageNamespaceIdentifier(), webPage.identifier(), topLevelOrigin, sessionStorageQuota());
+        it = sessionStorageNamespacesMap.set(topLevelOrigin.data(), WTFMove(sessionStorageNamespace)).iterator;
+    }
+    return it->value;
+}
+
+void WebStorageNamespaceProvider::copySessionStorageNamespace(WebCore::Page& srcPage, WebCore::Page& dstPage)
+{
+    ASSERT(sessionStorageQuota() != WebCore::StorageMap::noQuota);
+
+    const auto& srcWebPage = WebPage::fromCorePage(srcPage);
+    const auto& dstWebPage = WebPage::fromCorePage(dstPage);
+
+    auto srcNamespacesIt = m_sessionStorageNamespaces.find(srcWebPage.sessionStorageNamespaceIdentifier());
+    if (srcNamespacesIt == m_sessionStorageNamespaces.end())
+        return;
+
+    ASSERT(srcNamespacesIt->value.useCount);
+
+    auto& srcNamespacesMap = srcNamespacesIt->value.map;
+
+    auto& dstSessionStorageNamespaces = static_cast<WebStorageNamespaceProvider&>(dstPage.storageNamespaceProvider()).m_sessionStorageNamespaces;
+    auto dstNamespacesIt = dstSessionStorageNamespaces.find(dstWebPage.sessionStorageNamespaceIdentifier());
+    ASSERT(dstNamespacesIt != dstSessionStorageNamespaces.end());
+    ASSERT(dstNamespacesIt->value.useCount == 1);
+    auto& dstNamespacesMap = dstNamespacesIt->value.map;
+
+    if (auto networkProcessConnection = WebProcess::singleton().existingNetworkProcessConnection())
+        networkProcessConnection->connection().send(Messages::NetworkStorageManager::CloneSessionStorageNamespace(srcWebPage.sessionStorageNamespaceIdentifier(), dstWebPage.sessionStorageNamespaceIdentifier()), 0);
+
+    for (auto& [k, v] : srcNamespacesMap)
```

Comment:
Nit: maybe `[origin, namespace]` instead of `[k, v]`.
(I think we prefer using full name for clarity as in https://webkit.org/code-style-guidelines/.)

Reply:
Done